### PR TITLE
Optimize usages of "not"

### LINKED ISSUE #2793: EBPF backend generates incorrect C code for if (!table.hit)
Use these diffs for P4 code and then use p4c-ebpf to generate C code.

```C
diff --git a/testdata/p4_16_samples/switch_ebpf.p4 b/testdata/p4_16_samples/switch_ebpf.p4
index e08696d2d..2b1bcd370 100644
--- a/testdata/p4_16_samples/switch_ebpf.p4
+++ b/testdata/p4_16_samples/switch_ebpf.p4
@@ -81,13 +81,17 @@ control pipe(inout Headers_t hdr, out bool pass)
        Swap(false);
 
         pass = true;
-
+       if (!Check_src_ip.apply().hit) {
+           pass = false;
+       }
+/*
         switch (Check_src_ip.apply().action_run) {
         Reject: {
             pass = false;
         }
         NoAction: {}
         }
+*/
     }
 }
```

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
There are still a few reference outputs that I have to check.
This fixes #2793

### CODE REVIEW

**Comment by hesingh** [comment]
```
The following tests FAILED:
	1087 - p4/testdata/p4_16_samples/issue2546-1.p4 (Failed)
```

**Comment by hesingh** [comment]
```
./p4test --pp ./tmpbx4_2c12/issue2546-1.p4 --dump ./tmpbx4_2c12 --top4 FrontEndDump,MidEndLast,FrontEndLast --testJson --maxErrorCount 100 /p4c/testdata/p4_16_samples/issue2546-1.p4
Error compiling
def_use.h:379
Compiler Bug: no definitions found for 10 h/h.eth_hdr.eth_type
```

**Comment by mbudiu-vmw** [comment]
The change which triggered many reference outputs changes is converting `if (!a)` into `if (a)` by switching the if branches.

File: midend/removeMiss.cpp [hunk 1]
```diff
@@ -26,4 +26,19 @@ const IR::Node* DoRemoveMiss::preorder(IR::Member* expression) {
     return new IR::LNot(hit);
 }
 
+const IR::Node* DoRemoveMiss::preorder(IR::IfStatement* statement) {
+    if (!TableApplySolver::isMiss(statement->condition, refMap, typeMap))
```
Comment (fruffy) [inline midend/removeMiss.cpp:30]: Same here.

File: frontends/p4/simplify.cpp [hunk 2]
```diff
@@ -61,6 +61,16 @@ const IR::Node* DoSimplifyControlFlow::postorder(IR::BlockStatement* statement)
 
 const IR::Node* DoSimplifyControlFlow::postorder(IR::IfStatement* statement)  {
     LOG3("Visiting " << dbp(getOriginal()));
+    if (auto lnot = statement->condition->to<IR::LNot>()) {
+        // swap branches
+        statement->condition = lnot->expr;
+        auto e = statement->ifFalse;
+        if (!e)
+            e = new IR::EmptyStatement();
```
Comment (fruffy) [inline frontends/p4/simplify.cpp:69]: Shouldn't if statements have brackets?

File: midend/removeMiss.cpp [hunk 3]
```diff
@@ -26,4 +26,19 @@ const IR::Node* DoRemoveMiss::preorder(IR::Member* expression) {
...
+        return statement;
+    auto mem = statement->condition->to<IR::Member>();
+    CHECK_NULL(mem);
```
Comment (fruffy) [inline midend/removeMiss.cpp:33]: I have been fond of [checkedTo](https://github.com/p4lang/p4c/blob/08ef61a007b47b48945f86d60fc47b81b4d0fef1/ir/node.h#L64) for these.

File: lib/crash.cpp [hunk 4]
```diff
@@ -138,7 +138,8 @@ MTONLY(
         *p++ = "0123456789abcdef"[(a >> shift) & 0xf];
         shift -= 4; }
     *p++ = '\n';
-    (void) write(to_child, buffer, p-buffer);
+    auto _unused = write(to_child, buffer, p-buffer);
+    (void)_unused;
```
Comment (fruffy) [inline lib/crash.cpp:142]: To remove a warning?

File: test/gtest/load_ir_from_json.cpp [hunk 5]
```diff
@@ -49,7 +49,8 @@ TEST_F(FromJSONTest, load_ir_from_json) {
     ASSERT_FALSE(exitCode);
     exitCode = system("diff outputTO.json outputFROM.json");
     ASSERT_FALSE(exitCode);
-    system("rm -f outputFROM.json outputTo.json");
+    exitCode = system("rm -f outputFROM.json outputTo.json");
+    ASSERT_FALSE(exitCode);
```
Comment (fruffy) [inline test/gtest/load_ir_from_json.cpp:53]: To remove a warning?

**Review by fruffy — APPROVED** [review]

File: frontends/p4/simplify.cpp (hunk 2 above)
Comment (mbudiu-vmw) [reply frontends/p4/simplify.cpp:69]: No, our style guide does not require brakets for a single statement.
cpp lints decides here.

File: midend/removeMiss.cpp (hunk 3 above)
Comment (mbudiu-vmw) [reply midend/removeMiss.cpp:33]: Forgot about that one

File: lib/crash.cpp (hunk 4 above)
Comment (mbudiu-vmw) [reply lib/crash.cpp:142]: Yes, that is correct. Same for the next one.

**Pull request merged**